Add blob support

macintux opened the pull request:
The new `blob` type is effectively identical to `varchar` and is treated as such for most operations. They may diverge further as we look into proper utf-8 support for character strings, but that's a future discussion.

Since `blob` is basically `varchar` (and will be represented as such for downgrade purposes) rather than implement a new type, `riak_ql` now has the concept of "external" vs "internal" type representations. Externally a blob will be represented as a blob, while internally for storage it's a varchar.

Pull Requests:
* riak_kv https://github.com/basho/riak_kv/pull/1540
* riak_ql https://github.com/basho/riak_ql/pull/156
* riak_pb https://github.com/basho/riak_pb/pull/211
* riak_shell https://github.com/basho/riak_shell/pull/61
* riak_test https://github.com/basho/riak_test/pull/1212

macintux committed:
Add support for blobs

Make eleveldb happy by treating them as their underlying storage type:
varchar.

macintux committed:
Convert types where appropriate

* Review uses of field types in KV
* Use storage types for `do_insert`
* Correct type specs with `simple_field_type`

macintux committed:
Merge branch 'riak_ts-develop' of github.com:basho/riak_kv into feature_jrd_type-bifurcation

macintux committed:
Update type spec in include file

macintux committed:
Display blobs differently from varchar in explain

Untested, can't compile at the moment due to upstream changes, which I
can't comfortably merge without committing, so pffft.

macintux committed:
Merge branch 'riak_ts-develop' of github.com:basho/riak_kv into feature_jrd_type-bifurcation

macintux committed:
Syntax error

macintux committed:
Retrieve field types from the DDL module

My earlier code made the mistake of assuming all column names and types
were contained in the `'SELECT'` member. Instead, retrieve from the DDL
module, as any rational developer would have done the first time around.

macintux committed:
Make test failures easier to read

As I discovered while chasing an earlier bug, `?assertMatch` on large
data structures results in effectively unreadable errors. `?assertEqual`
gives better results, especially when combined with the `unite` library
for error formatting.

Thanks to Andy Till for this trick.

macintux committed:
Extend `explain` test to include blobs

macintux committed:
Drop extraneous line feed in hex output

andytill commented on src/riak_kv_ts_util.erl:
```diff
@@ -534,28 +534,38 @@ key_element_to_string(V) -> lists:flatten(io_lib:format("~p", [V])).
-element_to_quoted_string(V) when is_binary(V) -> varchar_quotes(V);
-element_to_quoted_string(V) when is_float(V) -> mochinum:digits(V);
-element_to_quoted_string(V) -> lists:flatten(io_lib:format("~p", [V])).
+element_to_quoted_string(V, blob) -> hex_as_string(hexlify(V));
```
The `mochihex` module is used in riak_ql. I used that instead of rolling my own so I didn't have to add unit tests to prove correctness.

macintux replied:
Fixed

andytill commented:
@borshop retry

macintux committed:
Drop custom hexlify in favor of mochihex

macintux commented:
This can be retried once `pb` and `ql` are merged.

macintux commented:
@borshop: retry

macintux committed:
Merge branch 'riak_ts-develop' of github.com:basho/riak_kv into feature_jrd_type-bifurcation

macintux commented:
Tests weren't running but needed another merge with upstream anyway. Running through tests manually.

macintux committed:
Fix stupid typo: qql -> ql

macintux committed:
Make `insert_test` work again

`riak_kv_compile_tab:insert_test` passes an incomplete DDL that did not
have enough information to survive the
`riak_ql_ddl:get_minimum_capability` gauntlet of checks.

By creating a single `riak_field_v1` record in the `fields` element, the
test can now proceed.

macintux committed:
Yet another incomplete DDL fix

macintux committed:
Yet another update for incomplete DDL tests

All `riak_kv` tests finally pass.

andytill commented:
+1 6601d25

andytill commented:
@borshop merge
